Swagger UI #25

svgogin opened the pull request:
* #25 Add swagger UI
1. Swagger UI added
2. Configuration work done
3. Some variables in the Exception handler were inlined
4. Missing parameters test added.

svgogin committed:
#25 Swagger UI (final)

1. OpenAPI specification added

alinkov commented on pom.xml:
```diff
@@ -70,6 +70,11 @@
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>io.springfox</groupId>
```
please use `springdoc-openapi-ui`

alinkov replied:
You can use https://www.baeldung.com/spring-rest-openapi-documentation as example

alinkov commented on pom.xml:
```diff
@@ -70,6 +70,11 @@
             <version>${testcontainers.version}</version>
             <scope>compile</scope>
```
please change it to test

alinkov commented on src/main/java/ru/svgogin/service/spark/swagger/SpringFoxConfig.java:
```diff
@@ -0,0 +1,68 @@
+package ru.svgogin.service.spark.swagger;
+
+import static org.testcontainers.shaded.com.google.common.collect.Lists.newArrayList;
```
don't use this dependency and also please fix this dependency scope.
It's better to use `List.of`

alinkov commented on src/main/java/ru/svgogin/service/spark/swagger/SpringFoxConfig.java:
```diff
@@ -0,0 +1,68 @@
+public class SpringFoxConfig {
+  @Bean
+  public Docket api() {
+    return new Docket(DocumentationType.SWAGGER_2)
```
please use 3 version

alinkov commented on src/main/java/ru/svgogin/service/spark/swagger/SpringFoxConfig.java:
```diff
@@ -0,0 +1,68 @@
+        .paths(PathSelectors.any())
+        .build().useDefaultResponseMessages(false)
+        .globalResponses(HttpMethod.GET, newArrayList(
+            new ResponseBuilder()
```
this is strange formatting.
Please use one of these:
```java
.globalResponses(HttpMethod.GET, List.of(
    new ResponseBuilder().code("500").description("InternalServerError").build(),
    new ResponseBuilder().code("400").description(ERROR003.label).build()
))
```
or
```java
.globalResponses(HttpMethod.GET, List.of(
    new ResponseBuilder()
        .code("500")
        .description("InternalServerError")
        .build(),
    new ResponseBuilder()
        .code("400")
        .description(ERROR003.label)
        .build()
))
```

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -87,10 +83,9 @@ private ErrorDto toErrorDto(ObjectError objectError) {
-    var errorCode = ErrorCode.ERROR003;
     var message = String.format("%s (%s)", violation.getMessageTemplate(),
         violation.getInvalidValue());
     log.warn(message);
```
can you move this logging into exception handling?

alinkov commented on src/main/java/ru/svgogin/service/spark/web/SparkController.java:
```diff
@@ -60,6 +334,61 @@ public SparkController(SparkService sparkService) {
+          content = {
+              @Content(
+                  examples = @ExampleObject(
+                      value = "{\n"
```
nit: you can update to java 17 and use text blocks

alinkov commented on src/main/java/ru/svgogin/service/spark/web/SparkController.java:
```diff
@@ -44,12 +146,184 @@ public SparkController(SparkService sparkService) {
+                  mediaType = "application/json",
+                  array = @ArraySchema(schema = @Schema(implementation = ErrorDto.class)))
+          }),
+      @ApiResponse(responseCode = "404",
```
you can create custom annotations and reuse them: (without `@ApiResponses`)
```java
@CompanyNotFoundResponse
```

```java
@Target({PARAMETER, METHOD, FIELD, ANNOTATION_TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Inherited
@ApiResponse(responseCode = "404", description = "Company not found")
public @interface CompanyNotFoundResponse {
}
```

svgogin committed:
#25 Swagger UI

1. Logging moved into the exception handlers
2. Project updated to java 17
3. Text blocks added
4. Annotations added

svgogin committed:
#25 Swagger UI

5. JDK version changed to 17 in maven.yml

alinkov commented on src/main/java/ru/svgogin/service/spark/apiresponses/AllCompaniesFoundResponse.java:
```diff
@@ -0,0 +1,56 @@
+package ru.svgogin.service.spark.apiresponses;
```
nit: I think you can rename package to `api`

alinkov commented on src/main/java/ru/svgogin/service/spark/apiresponses/InvalidPathAndBodyParamsResponse.java:
```diff
@@ -0,0 +1,52 @@
+                            "code": "ERROR003",
+                            "message": "InvalidRequestBodyFormat (ogrn)"
+                        },
+                        {        "code": "ERROR003",
```
```suggestion
                        {       
                               "code": "ERROR003",
```

alinkov commented on src/main/java/ru/svgogin/service/spark/apiresponses/InvalidPathAndBodyParamsResponse.java:
```diff
@@ -0,0 +1,52 @@
+                        {        "code": "ERROR003",
+                            "message": "InvalidRequestBodyFormat (inn)"
+                        },
+                        {        "code": "ERROR004",
```
nit: new line

alinkov commented on src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:
```diff
@@ -62,7 +63,7 @@
     Iterable<ErrorDto> errors = ex.getBindingResult().getAllErrors().stream()
         .map(this::toErrorDto).collect(Collectors.toList());
-
+    errors.forEach(error -> log.warn(error.getMessage()));
```
I think you can log full error. It can help you:
```java
ex.getBindingResult().getAllErrors().forEach(log::warn);
```

svgogin committed:
#25 Swagger UI

1. Logging in RestResponseEntityExceptionHandler improved.
2. Package apiresponses renamed to api.
3. Testcontainers.version updated to 1.16.3
4. Code style improved.

svgogin committed:
#25 Swagger UI

1. Testcontainers scope changed from compile to test
